heartbeat.go

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2018 Snowflake Computing Inc. All right reserved.
+
+package gosnowflake
+
+import (
+	"fmt"
+	"net/url"
+	"time"
+
+	"context"
+
+	"io/ioutil"
+	"net/http"
+
+	"encoding/json"
+
+	"github.com/satori/go.uuid"
+)
+
+const (
+	// One hour interval should be good enough to renew tokens for four hours master token validity
+	heartBeatInterval = 60 * time.Minute
+)
+
+type heartbeat struct {
+	restful      *snowflakeRestful
+	ShutdownChan chan int
+}
+
+func (hc *heartbeat) run() {
+	hbTimer := time.NewTimer(heartBeatInterval)
+	for {
+		go func() {
+			<-hbTimer.C
+			err := hc.heartbeatMain()
+			if err != nil {
+				glog.V(2).Info("failed to heartbeat")
+				return
+			}
+			hbTimer = time.NewTimer(heartBeatInterval)
+		}()
+		select {
+		case <-hc.ShutdownChan:
+			// stop
+			hbTimer.Stop()
+			glog.V(2).Info("stopping")
+			return
+		default:
+			// no more download
+			glog.V(2).Info("no shutdown signal")
+		}
+		time.Sleep(1 * time.Second)
+	}
+}
```

Comment:
main routine. Not sure if this way is optimal. Any suggestion would be appreciated.